File: w6/src/maxflow.rs
use std::{
    collections::{HashMap, HashSet, VecDeque},
    io::{self, Read},
};

#[derive(Debug, PartialEq, PartialOrd, Clone, Ord, Eq)]
struct Edge {
    capacity: i64,
    flow: i64,
}

impl Edge {
    fn get_remainging_flow(&self) -> i64 {
        return self.capacity - self.flow;
    }
}

pub fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace();

    let n: usize = input.next().unwrap().parse().unwrap();
    let m: usize = input.next().unwrap().parse().unwrap();
    let s: usize = input.next().unwrap().parse().unwrap();
    let t: usize = input.next().unwrap().parse().unwrap();

    let mut graph: Vec<HashMap<usize, Edge>> = vec![HashMap::new(); n];
    let mut max_cap = 0;
    for _ in 0..m {
        let u: usize = input.next().unwrap().parse().unwrap();
        let v: usize = input.next().unwrap().parse().unwrap();
        let c: i64 = input.next().unwrap().parse().unwrap();
        if c > max_cap {
            max_cap = c;
        }
        graph[u].insert(
            v,
            Edge {
                capacity: c,
                flow: 0,
            },
        );
    }
    let (max_flow, _) = solve_flow(&mut graph, s, t, max_cap);
    let mut max_flow_edges: HashSet<(usize, usize, i64)> = HashSet::new();
    for from_node in 0..graph.len() {
        for (to_node, edge) in &graph[from_node] {
            if edge.flow > 0 && edge.capacity > 0 {
                max_flow_edges.insert((from_node, *to_node, edge.flow));
            }
        }
    }

    println!("{} {} {}", n, max_flow, max_flow_edges.len());

    for line in max_flow_edges {
        println!("{} {} {}", line.0, line.1, line.2);
    }
}

fn bfs(
    graph: &Vec<HashMap<usize, Edge>>,
    source: usize,
    sink: usize,
    cap: i64,
) -> Option<(i64, Vec<(usize, usize)>)> {
    let mut parents: HashMap<usize, usize> = HashMap::new();
    let mut queue: VecDeque<usize> = VecDeque::from([source]);
    while !queue.is_empty() {
        let from_node = queue.pop_front().unwrap();
        for (to_node, edge) in &graph[from_node] {
            if edge.get_remainging_flow() >= cap && !parents.contains_key(&to_node) {
                parents.insert(*to_node, from_node);
                queue.push_back(*to_node);
                if *to_node == sink {
                    let mut path: Vec<(usize, usize)> = vec![];
                    let mut curr_node = sink;
                    let mut bottle_neck = edge.get_remainging_flow();
                    while curr_node != source {
                        let tmp = *parents.get(&curr_node).unwrap();
                        bottle_neck = bottle_neck
                            .min(graph[tmp].get(&curr_node).unwrap().get_remainging_flow());
                        path.push((tmp, curr_node));
                        curr_node = tmp;
                    }
                    return Some((bottle_neck, path));
                }
            }
        }
    }
    return None;
}

fn solve_flow(
    graph: &mut Vec<HashMap<usize, Edge>>,
    source: usize,
    sink: usize,
    max_cap: i64,
) -> (i64, Vec<(usize, usize)>) {
    let mut curr_cap = max_cap;
    let mut max_flow = 0;
    let mut path: Vec<(usize, usize)> = vec![];
    loop {
        match bfs(&graph, source, sink, curr_cap) {
            Some((bottle_neck, found_path)) => {
                max_flow += bottle_neck;
                for (from_node, to_node) in &found_path {
                    graph[*from_node].get_mut(&to_node).unwrap().flow += bottle_neck;
                    match graph[*to_node].get_mut(&from_node) {
                        Some(edge) => {
                            edge.flow -= bottle_neck;
                        }
                        None => {
                            graph[*to_node].insert(
                                *from_node,
                                Edge {
                                    capacity: 0,
                                    flow: -bottle_neck,
                                },
                            );
                        }
                    }
                }
                path = found_path;
            }
            None => {
                if curr_cap > 0 {
                    curr_cap /= 2;
                }
                if curr_cap < 1 {
                    return (max_flow, path);
                }
            }
        }
    }
}

File: w6/src/maxflow.py
from collections import defaultdict

graph = defaultdict(list)

visited = set()

def scaling(current, cap, visited):
    if current in visited:
        return False
    if current == t:
        return True
    
    visited.add(current)
    
    for adjacent in graph[current]:
        print("adjacent: ")
        print(adjacent)
        if adjacent[1] >= cap:
            if scaling(adjacent[0],cap, visited):
                adjacent[1] -= cap
                adjacent[2] += cap
                return True
            
    return False


n,m,s,t = map(int, input().split())
# n = number of nodes in graph
# m = number of edges in graph
# s = source
# t = sink

queue = []
maxcap = 0
# [[to,capacity,used capacity]]
for i in range(m):
    u,v,c = map(int, input().split())
    graph[u].append([v,c,0])
    if c > maxcap:
        maxcap = c

threshold = 2**27
while(maxcap < threshold):
    threshold = threshold//2

done = False
while(not done):
    visited = set()
    if scaling(s,threshold, visited):
        continue
    threshold = threshold//2
    if threshold == 0:
        threshold = 1
    visited = set()
    if threshold == 1 and not scaling(s,threshold, visited):
        done = True
    continue

maxflow = 0
for edge in graph[s]:
    maxflow += edge[2]

counter = 0
edgesused = set()
for key,value in graph.items():
    for e in value:
        if e[2] != 0:
            counter += 1
            edgesused.add((key,e[0],e[2]))


print(f"{n} {maxflow} {counter}")
for edge in sorted(edgesused):
    print(*edge)